nntrainer/compiler/multiout_realizer.cpp

```diff
@@ -9,16 +9,99 @@
  * @author Jihoon Lee <jhoon.it.lee@samsung.com>
  * @bug No known bugs except for NYI items
  */
+#include <stdexcept>
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include <common_properties.h>
 #include <compiler_fwd.h>
+#include <layer_node.h>
 #include <multiout_realizer.h>
+#include <remap_realizer.h>
 
 namespace nntrainer {
 MultioutRealizer::~MultioutRealizer() {}
 
 GraphRepresentation
 MultioutRealizer::realize(const GraphRepresentation &reference) {
+  GraphRepresentation processed(reference.begin(), reference.end());
+
+  std::unordered_multiset<std::string> freq_map;
+  std::unordered_set<std::string> node_names;
+  std::unordered_set<std::string> connection_names;
+
+  /// 1. build frequency map and connection names
+  for (auto &node : reference) {
+    NNTR_THROW_IF(node_names.count(node->getName()), std::invalid_argument)
+      << "node name clashes: " << node->getName();
+    node_names.emplace(node->getName());
+
+    for (unsigned int i = 0, num_nodes = node->getNumInputConnections();
+         i < num_nodes; ++i) {
+      props::InputConnection c(props::Connection(
+        node->getInputConnectionName(i), node->getInputConnectionIndex(i)));
+      auto uniq_name = to_string(c);
+      connection_names.emplace(uniq_name);
+      freq_map.emplace(uniq_name);
+    }
+  }
+
+  /// 2. for each connection names, if a connection is referenced multiple
+  /// times, create multioutput node and remap to multi output node index
+  std::unordered_multimap<std::string /**< original id */,
+                          std::shared_ptr<LayerNode> /**< created node */>
+    multiout_nodes;
+
+  for (auto &con_name : connection_names) {
+    props::InputConnection con;
+    from_string(con_name, con);
+    if (freq_map.count(con_name) <= 1) {
+      continue;
+    }
+
+    std::string id = con.get().getName();
+    auto idx = con.get().getIndex();
+
+    std::stringstream ss;
+    /// {connection_name}/generated_out_{index}
+    ss << id << "/generated_out_" << idx;
+    while (node_names.count(ss.str()) != 0) {
```

Comment:
layer_name will be made unique if the name is already taken. IMO this can be skipped.

Reply:
This is checking if `{connection_name}/generated_out_{index}` is already taken from the given representation. At this stage  we need to ensure the name is unique else remapping will be wrong.

consider below ini.

```ini
[fc]
type=input
input_shape=3

[fc/generated_out_0]
type=input
input_shape=1

[a]
input_layers=fc

[b]
input_layers=fc
```
will be turned to below without this check.
```ini
[fc]
type=input
input_shape=3

[fc/generated_out_0]
type=input
input_shape=1

[fc/generated_out_0] 
type=multiout
input_layers=fc(0)

[a]
input_layers=fc/generated_out_0

[b]
input_layers=fc/generated_out_0
```

Reply:
yeah, its better to have same semantics everywhere. So, renaming here is fine then.

FYI, in this case, the list of layers is not representative of the ini. 
In ini, two layers with same name gets interpreted by the ini file format as redefinition of the same layer.
But in the layer list, as layer is already defined, the layer name will automatically be updated.

Reply:
Yeah, right.

What we can guarantee is the exact ini is always interpreted the same way